engine/robot.py

```diff
@@ -330,8 +374,97 @@ def set_phase(self, new_phase):
+            if (self.phase == Phase.TRAVERSE) or (self.phase == Phase.RETURN) or (self.phase == Phase.DOCKING) or (
+                    self.phase == Phase.AVOID_OBSTACLE):
+                if curr_ultrasonic_value < self.detect_obstacle_range:
+                    # didn't check whether we can reach goal before contacting obstacle because obstacle detection does
+                    # not detect angle, so obstacle could be calculated to be falsely farther away than the goal. Not
+                    # optimal because in cases, robot will execute boundary following when it can reach goal
+                    self.dist_to_goal = math.sqrt(
+                        (self.state[0] - self.goal_location[0]) ** 2 + (self.state[1] - self.goal_location[1]) ** 2)
+                    self.avoid_obstacle = True
+                    with open(ROOT_DIR + '/tests/functionality_tests/csv/avoid_obstacle_result.csv', 'a') as fd:
+                        fd.write("Avoid" + '\n')
+                else:
+                    self.avoid_obstacle = False
+                    with open(ROOT_DIR + '/tests/functionality_tests/csv/avoid_obstacle_result.csv', 'a') as fd:
+                        fd.write("Not Avoid" + '\n')
+            if curr_ultrasonic_value < 0:
+                self.set_phase(Phase.FAULT)  # value should not go below 0; sensor is broken
+                with open(ROOT_DIR + '/tests/functionality_tests/csv/avoid_obstacle_result.csv', 'a') as fd:
+                    fd.write("Fault" + '\n')
+
+        # time.sleep(10)  # don't hog the cpu
+
+    def execute_avoid_obstacle(self, dist_to_goal, prev_phase, init_threshold, goal_threshold):
+        """ Execute obstacle avoidance
+            Args:
+                dist_to_goal (Double): The distance from the robot to the goal at the start of the phase
+                prev_phase (Phase): The previous phase to return to after executing obstacle avoidance
+                init_threshold (Double): Radius from initial position that robot has to leave before it can detect if
+                    goal is unreachable
+                goal_threshold (Double): Threshold from goal that will be detected as reaching goal
+            Returns:
+                prev_phase (Phase)
+        """
+        init_x = self.x_pos
+        init_y = self.y_pos
+        # threshold is arbitrary, set later. sometimes location reading will be inaccurate or not frequent enough to
+        # read that we've arrived back to the init x y pos. make it small enough that robot actually leaves threshold
+        # at some time during boundary following or add timeout to branch making gate = True
+        gate = False
+        dist_to_goal_decreases = False
+        boundary_traversed = False
+        while True:
+            if not self.fault:  # fault has priority over obstacle avoidance
+                dist_from_init = math.sqrt((self.x_pos - init_x) ** 2 + (self.y_pos - init_y) ** 2)
```

Comment:
maybe we should make a general distance helper function + also use in track_obstacle